### ISSUE TITLE
Reimplement MP-CNN model in PyTorch

### ISSUE THREAD

**Issue by lintool** [castorini/Castor#15]
Now that we have https://github.com/castorini/MP-CNN-Torch integrated into Castorini - it would be nice to have a PyTorch implementation of the model.

@gauravbaruah it occurs to me that MP-CNN does not have any of the issues associated with SM re: idf feature. re: #8

**Comment by tuzhucheng** [comment]
I'll take this task.

**Comment by lintool** [comment]
Hi @tuzhucheng - while you are working on this port, can you please look at the following:

There are a few aspects of Hua's EMNLP model that isn't symmetric:

+ See Figure 3. Building Block A does max/min/mean pooling, while Building Block B only does max/min pooling. Add mean pooling to B.
+ For Algorithm 1 (horizontal comparison), it doesn't compare on Building Block B. Add that.
+ Alg 1 and Alg 2 use different similarity metrics (i.e., comU2 in Alg1 vs. comU1 in Alg2); use the same in both.

IIRC, the model right now isn't symmetric because Hua didn't have time to run a complete set of experiments... but we should address these issues if we're going to re-implement anyway.

**Comment by tuzhucheng** [comment]
Sounds good!

**Comment by tuzhucheng** [comment]
Just an update on this issue, I've re-implemented the original MP-CNN paper. And getting these results on the SICK dataset compared with the results reported in the paper.

| Implementation        | Pearson's r           | Spearman's rho  |
| ------------- |:-------------:| -----:|
| Torch      | 0.8686 | 0.8047 |
| PyTorch      | 0.8154      |   0.7328 |

There are several variations in the current implementation that might be responsible differences:
* The PyTorch implementation uses the KLDivLoss loss function, which is a built-in class. The paper/Torch implementation uses KL-divergence loss plus an L2 regularizer. Need to code a custom loss function in PyTorch to include L2 regularization.
* PyTorch implementation uses batched training to speed up the training process (~40 minutes/epoch to <2 minutes/epoch), which means shorter sentence embeddings need to be 0-padded at the end. The Torch implementation does not do batching hence there is no 0-padding.
* For the "element-wise absolute difference measure magnitude differences" in the comparison unit, maybe I misinterpreted this and instead uses the sum of the absolute differences. Will check how the Torch Lua code does this.
* Different treatment of unknown vectors - The PyTorch implementation uses `normal(0, 1)` following how [torchtext](https://github.com/pytorch/text/blob/815dfe9ddc87502bc02d7b791502cafdad96c758/torchtext/vocab.py#L166) does it while the Torch implementation uses `uniform(-0.05, 0.05)`.

There are probably other differences too. Will make some tweaks and see again.

**Comment by hohoCode** [comment]
Sometime L2 loss may not be necessary since it could be too strong.

**Comment by tuzhucheng** [comment]
Yep I actually meant mini-batch :)

Here are the latest numbers for SICK. As a part of #30, going to try to use a integration framework to ensure these numbers are replicable.

Implementation | Pearson's r | Spearman's rho
-- | -- | --
Torch | 0.8686 | 0.8047
PyTorch | 0.868768 | 0.807172

Currently trying to adjust hyperparameters for training MSRVID properly.

**Comment by tuzhucheng** [comment]
Following up on this from a while ago:

>>> There are a few aspects of Hua's EMNLP model that isn't symmetric:

>>> * See Figure 3. Building Block A does max/min/mean pooling, while Building Block B only does max/min pooling. Add mean pooling to B.
>>> * For Algorithm 1 (horizontal comparison), it doesn't compare on Building Block B. Add that.
>>> * Alg 1 and Alg 2 use different similarity metrics (i.e., comU2 in Alg1 vs. comU1 in Alg2); use the same in both.

I've done experiments on the first and third points on the SICK dataset. For the second point, the comparison for Building Block B is done in the second part of Alg 2 using Comp U1, which is a superset of Comp U2. Hence I don't believe testing the second point is necessary besides comparing the performance of using different comparison units (point 3).

| Version        | Pearson's r           | Spearman's rho  |
| ------------- |:-------------:| -----:|
| Base MP-CNN      | 0.8751 | 0.8186 |
| Symmetrical Pooling - Add mean pooling to B     | 0.8755      |   0.8194 |
| Use Comp U1 for all comparison algorithms | 0.8733      |    0.8170 |
| Use Comp U2 for all comparison algorithms | 0.8794      |    0.8227 |

All reported values are averaged among results obtained using 5 random seeds (fixed between different versions).

Adding mean pooling improved performance negligibly. Using Comp U1 worsened performance while using Comp U2 improved performance by almost half a point. Comp U2, unlike Comp U1, does not have element-wise absolute difference measure magnitude differences - which significantly reduces the number of parameters of the model.

**Comment by tuzhucheng** [comment]
Here are the results for MSRVID

| Version        | Dev Pearson's r           | Test Pearson's r  |
| ------------- |:-------------:| -----:|
| Base MP-CNN      | 0.9812  | 0.9031 |
| Symmetrical Pooling - Add mean pooling to B      | 0.9943  |  0.9062  |
| Use Comp U1 for all comparison algorithms | 0.9517  | 0.9093  |
| Use Comp U2 for all comparison algorithms | 0.9917  |  0.9100 |

MSRVID doesn't have a dev set, so dev is ~20% of the training set used for model selection.

**Comment by tuzhucheng** [comment]
On both of these datasets, adding mean pooling to Block B and using U2 individually helped with performance. I'll run some experiments on trying both of those together.

**Comment by tuzhucheng** [comment]
Here are results combining using only Comp U2 and mean pooling, which individually helped improved performance. It turns out using both of them doesn't do better than using just Comp U2 for all comparison algorithms.

SICK

Version | Pearson's r | Spearman's rho
-- | -- | --
Base MP-CNN | 0.8751 | 0.8186
Only use mean pooling for Block B | 0.8755 | 0.8194
Only use Comp U2 for all comparison algorithms | 0.8794 | 0.8227
Using both mean pooling and Comp U2 | 0.8797 | 0.8225

MSRVID

Version | Dev Pearson's r | Test Pearson's r
-- | -- | --
Base MP-CNN | 0.9812 | 0.9031
Only use mean pooling for Block B | 0.9943 | 0.9062
Only use Comp U2 for all comparison algorithms | 0.9917 | 0.9100
Using both mean pooling and Comp U2 | 0.9798 | 0.9070

**Comment by lintool** [comment]
Next interesting question - what if we removed mean pooling for both block A and B? Try: max pooling only, max and min pooling.

**Comment by tuzhucheng** [comment]
Here are the results using max pooling only, and max and min pooling. Compared to the base version, removing mean pooling either helped performance or worsened it by a tiny amount. On one dataset having max and min pooling was better than than just max pooling, but on the other it was not.

SICK

Version | Pearson's r | Spearman's rho
-- | -- | --
Base MP-CNN | 0.8751 | 0.8186
Max Pooling Only | 0.8739 | 0.8160
Max Pooling and Min Pooling (No Mean) | 0.8756 | 0.8198

MSRVID

Version | Dev Pearson's r | Test Pearson's r
-- | -- | --
Base MP-CNN | 0.9812 | 0.9031
Max Pooling Only | 0.9840 | 0.9045
Max Pooling and Min Pooling (No Mean) | 0.9822 | 0.9027
